web_monitoring/db.py

```diff
@@ -1,47 +1,608 @@
+              'capture_time': time_range_string(start_date, end_date)}
+    if source_metadata is not None:
+        for k, v in source_metadata.items():
+            params[f'source_metadata[{k}]'] = v
+    url = f'{api_url()}/pages/{page_id}/versions'
+    res = requests.get(url, auth=auth(), params=params)
+    res.raise_for_status()
+    result = res.json()
+    # In place, replace datetime strings with datetime objects.
+    for v in result['data']:
+        v['created_at'] = parse_timestamp(v['created_at'])
+        v['updated_at'] = parse_timestamp(v['updated_at'])
+        v['capture_time'] = parse_timestamp(v['capture_time'])
+    return result
+
+
+@ensure_credentials
+def list_versions(chunk=None, chunk_size=None,
+                  start_date=None, end_date=None,
+                  source_type=None, version_hash=None,
+                  source_metadata=None):
+    """
+    List all Versions, optionally filtered by serach criteria.
+
+    Parameters
+    ----------
+    chunk : integer, optional
+        pagination parameter
+    chunk_size : integer, optional
+        number of items per chunk
+    start_date : datetime, optional
+    end_date : datetime, optional
+    source_type : string, optional
+    version_hash : string, optional
+    source_metadata : dict, optional
+        Examples:
+
+        * ``{'version_id': 12345678}``
+        * ``{'account': 'versionista1', 'has_content': True}``
+
+    Returns
+    -------
+    response : dict
+    """
+    params = {'chunk': chunk,
+              'chunk_size': chunk_size,
+              'capture_time': time_range_string(start_date, end_date),
+              'source_type': source_type,
+              'version_hash': version_hash}
+    if source_metadata is not None:
+        for k, v in source_metadata.items():
+            params[f'source_metadata[{k}]'] = v
+    url = f'{api_url()}/versions'
+    res = requests.get(url, auth=auth(), params=params)
+    res.raise_for_status()
+    result = res.json()
+    # In place, replace datetime strings with datetime objects.
+    for v in result['data']:
+        v['created_at'] = parse_timestamp(v['created_at'])
+        v['updated_at'] = parse_timestamp(v['updated_at'])
+        v['capture_time'] = parse_timestamp(v['capture_time'])
+    return result
+
+
+@ensure_credentials
+def get_version(version_id):
+    """
+    Lookup a specific Version by ID.
+
+    Parameters
+    ----------
+    version_id : string
+
+    Returns
+    -------
+    response : dict
+    """
+    url = f'{api_url()}/versions/{version_id}'
+    res = requests.get(url, auth=auth())
+    res.raise_for_status()
+    result = res.json()
+    data = result['data']
+    data['capture_time'] = parse_timestamp(data['capture_time'])
+    data['updated_at'] = parse_timestamp(data['updated_at'])
+    data['created_at'] = parse_timestamp(data['created_at'])
+    return result
+
+
+@ensure_credentials
+def post_version(page_id, uuid, capture_time, uri, version_hash,
```

Comment:
Since you aren’t going by HTTP verbs for lists, would `submit` or `create` or `add` be clearer here?

Reply:
Makes sense. (Would you find HTTP verbs much more clear? My motivation here was distinguish between functions that give a list-like result for `data` as opposed to a single object.)

Reply:
Nah, I don’t think HTTP verbs would necessarily be more clear. I generally agree that names that distinguish the kind of action taken/content returned are better.

(In that sense, the `list_*/get_*` names are great. I was mostly reacting to the fact that the name here is `post_*`, but the first line of the docstring is `Submit a…`, implying that “post” was not really the clearest verb.)